File: Array/zhangjunwei/queue/MaxQueue_6_18.java
package com.lanfairy.ndkstudy;

import java.util.Iterator;
import java.util.LinkedList;

public class MaxQueue {
    private LinkedList<Integer> items;
    private int maxItem;

    public MaxQueue() {
        items = new LinkedList<>();
        maxItem = -1;
    }

    private boolean isEmpty(){
        return items.isEmpty();
    }

    public int max_value() {
        return maxItem;
    }

    public void push_back(int value) {
        items.offer(value);
        if (value>maxItem)
            maxItem = value;
    }

    public int pop_front() {
        if (items.isEmpty())return -1;
        int res = items.poll();
        if (items.isEmpty()){
            maxItem = -1;
            return res;
        }

        Iterator<Integer> iterator = items.iterator();
        maxItem = -1;
        while (iterator.hasNext()){
            Integer value = iterator.next();
            if (value>maxItem)
                maxItem = value;
        }

        return res;
    }


}

File: Array/zhangjunwei/queue/MaxQueue_6_18.kt
package mykotlin

import java.util.*

class MaxQueue() {
    private var queue: LinkedList<Int> = LinkedList()
    private var maxItem: Int = -1
    fun max_value(): Int {
        return maxItem
    }

    fun push_back(value: Int) {
        queue.offer(value)
        if (maxItem < value)
            maxItem = value
    }

    fun pop_front(): Int {
        if (queue.isEmpty()) {
            maxItem = -1
            return -1
        }
        var res = queue.poll()

        if (queue.isEmpty()){
            maxItem = -1
            return -1
        }
        if (res == maxItem) {
            maxItem = queue.peek()
            var size = queue.size
            while (size>0){
                this.push_back(queue.poll())
                size--
            }
        }
        return res
    }

}

/**
 * Your MaxQueue object will be instantiated and called as such:
 * var obj = MaxQueue()
 * var param_1 = obj.max_value()
 * obj.push_back(value)
 * var param_3 = obj.pop_front()
 */